File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pow/PoWMiner.java
package org.tdf.sunflower.consensus.pow;

import lombok.extern.slf4j.Slf4j;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.ByteUtil;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.consensus.AbstractMiner;
import org.tdf.sunflower.events.NewBestBlock;
import org.tdf.sunflower.events.NewBlockMined;
import org.tdf.sunflower.facade.RepositoryReader;
import org.tdf.sunflower.facade.TransactionPool;
import org.tdf.sunflower.state.Address;
import org.tdf.sunflower.types.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.Random;
import java.util.concurrent.*;

import static org.tdf.sunflower.ApplicationConstants.MAX_SHUTDOWN_WAITING;
import static org.tdf.sunflower.types.Transaction.ADDRESS_LENGTH;

@Slf4j(topic = "pow-miner")
public class PoWMiner extends AbstractMiner {
    private final TransactionPool transactionPool;
    private final ConsensusConfig config;
    private final PoW poW;
    private final ForkJoinPool threadPool = (ForkJoinPool) Executors.newWorkStealingPool();
    private ScheduledExecutorService minerExecutor;
    private volatile boolean stopped;

    private volatile long currentMiningHeight;
    private volatile boolean working;
    private Future<?> task;

    public PoWMiner(
        ConsensusConfig config,
        TransactionPool tp,
        PoW pow
    ) {
        super(pow.getAccountTrie(), pow.getEventBus(), config, pow.getTransactionPool());
        this.transactionPool = tp;
        this.config = config;
        this.poW = pow;

        getEventBus().subscribe(NewBestBlock.class, e -> {
            if (e.getBlock().getHeight() >= this.currentMiningHeight) {
                // cancel mining
                this.working = false;
                this.currentMiningHeight = 0;
            }
        });
    }

    @Override
    protected Transaction createCoinBase(long height) {
        return Transaction.builder()
            .gasLimit(ByteUtil.EMPTY_BYTE_ARRAY)
            .receiveAddress(Address.empty().getBytes())
            .data(PoWBios.UPDATE.encode())
            .value(ByteUtil.EMPTY_BYTE_ARRAY)
            .gasPrice(ByteUtil.EMPTY_BYTE_ARRAY)
            .nonce(ByteUtil.longToBytesNoLeadZeroes(height))
            .build();
    }

    @Override
    protected Header createHeader(Block parent) {
        return Header.builder()
            .hashPrev(parent.getHash())
            .coinbase(config.getMinerCoinBase())
            .createdAt(System.currentTimeMillis() / 1000)
            .height(parent.getHeight() + 1)
            .build();
    }

    @Override
    protected boolean finalizeBlock(Block parent, Block block) {
        Uint256 nbits = poW.bios.getNBits(parent.getHash());
        Random rd = new Random();
        byte[] nonce = new byte[8];
        rd.nextBytes(nonce);
        log.info("start finish pow target = {}", nbits.getDataHex());
        this.working = true;
        while (PoW.compare(PoW.getPoWHash(block), nbits.getData()) > 0) {
            if (!working) {
                log.info("mining canceled");
                return false;
            }
            rd.nextBytes(nonce);
            block.setNonce(HexBytes.fromBytes(Arrays.copyOfRange(nonce, 0, nonce.length)));
        }
        log.info("pow success");
        this.working = false;
        return true;
    }

    @Override
    public void start() {
        this.stopped = false;
        minerExecutor = Executors.newSingleThreadScheduledExecutor();
        minerExecutor.scheduleAtFixedRate(() -> {
            try {
                this.tryMine();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }, 0, Math.max(1, config.getBlockInterval() / 4), TimeUnit.SECONDS);
    }

    @Override
    public void stop() {
        if (stopped) return;
        this.working = false;
        minerExecutor.shutdown();
        try {
            minerExecutor.awaitTermination(MAX_SHUTDOWN_WAITING, TimeUnit.SECONDS);
            threadPool.awaitTermination(MAX_SHUTDOWN_WAITING, TimeUnit.SECONDS);
            log.info("miner stopped normally");
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        this.task = null;
        stopped = true;
    }

    @Override
    public HexBytes getMinerAddress() {
        return config.getMinerCoinBase();
    }

    public void tryMine() {
        if (!config.enableMining() || stopped) {
            return;
        }
        if (config.getMinerCoinBase() == null || config.getMinerCoinBase().size() != ADDRESS_LENGTH) {
            log.warn("pow miner: invalid coinbase address {}", config.getMinerCoinBase());
            return;
        }
        if (working || task != null) return;

        try (RepositoryReader rd = poW.getRepo().getReader()) {
            Block best = rd.getBestBlock();
            log.debug("try to mining at height " + (best.getHeight() + 1));
            this.currentMiningHeight = best.getHeight() + 1;
            long current = System.currentTimeMillis() / 1000;
            if (current <= best.getCreatedAt())
                return;
        }

        Runnable task = () -> {
            try (RepositoryReader rd = poW.getRepo().getReader()) {
                BlockCreateResult res = createBlock(
                    rd.getBestBlock(),
                    Collections.emptyMap());
                if (res.getBlock() != null) {
                    log.info("mining success block: {}", res.getBlock().getHeader());
                }
                getEventBus().publish(new NewBlockMined(res.getBlock(), res.getInfos()));
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                this.task = null;
            }
        };
        this.task = threadPool.submit(task);
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pow/PoWMiner.kt
package org.tdf.sunflower.consensus.pow

import org.slf4j.LoggerFactory
import org.tdf.common.types.Constants.NONCE_SIZE
import org.tdf.common.types.Uint256
import org.tdf.common.util.hex
import org.tdf.sunflower.ApplicationConstants.MAX_SHUTDOWN_WAITING
import org.tdf.sunflower.consensus.AbstractMiner
import org.tdf.sunflower.events.NewBestBlock
import org.tdf.sunflower.events.NewBlockMined
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.facade.TransactionPool
import org.tdf.sunflower.types.*
import java.util.*
import java.util.concurrent.*

class PoWMiner(
    private val config: ConsensusConfig,
    private val transactionPool: TransactionPool,
    private val poW: PoW
) : AbstractMiner(poW.accountTrie, poW.eventBus, config, poW.transactionPool) {
    private val threadPool = Executors.newWorkStealingPool() as ForkJoinPool
    private lateinit var minerExecutor: ScheduledExecutorService

    override val chainId: Int
        get() = config.chainId

    @Volatile
    private var stopped = false

    @Volatile
    private var currentMiningHeight: Long = 0

    @Volatile
    private var task: Future<*>? = null

    override fun createCoinBase(height: Long): Transaction {
        return Transaction(
            data = PoWBios.UPDATE.encode().hex(),
            nonce = height
        )
    }

    override fun createHeader(parent: Block, createdAt: Long): Header {
        return HeaderImpl(
            hashPrev = parent.hash,
            coinbase = config.coinbase!!,
            createdAt = createdAt,
            height = parent.height + 1
        )
    }

    override fun finalizeBlock(rd: RepositoryReader, parent: Block, block: Block): Block {
        var nbits: Uint256
        var b = block
        poW.repo.reader.use { rd -> nbits = poW.bios.getNBits(rd, parent.hash) }
        val rd = Random()
        val nonce = ByteArray(NONCE_SIZE)
        log.info("start finish pow target = {}", nbits.byte32.hex())
        while (PoW.compare(PoW.getPoWHash(b), nbits.byte32) > 0) {
            rd.nextBytes(nonce)
            b = b.copy(header = b.header.impl.copy(nonce = nonce.hex()))
        }
        log.info("pow success")
        return b
    }

    override fun start() {
        stopped = false
        minerExecutor = Executors.newSingleThreadScheduledExecutor()
        minerExecutor.scheduleAtFixedRate({
            try {
                tryMine()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }, 0, Math.max(1, config.blockInterval / 4).toLong(), TimeUnit.SECONDS)
    }

    override fun stop() {
        if (stopped) return
        minerExecutor!!.shutdown()
        try {
            minerExecutor!!.awaitTermination(MAX_SHUTDOWN_WAITING, TimeUnit.SECONDS)
            threadPool.awaitTermination(MAX_SHUTDOWN_WAITING, TimeUnit.SECONDS)
            log.info("miner stopped normally")
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        task = null
        stopped = true
    }

    fun tryMine() {
        if (!config.enableMining || stopped) {
            return
        }
        if (config.coinbase == null) {
            log.warn("pow miner: invalid coinbase address {}", config.coinbase)
            return
        }
        if (task != null) return
        poW.repo.reader.use { rd ->
            val best = rd.bestBlock
            log.debug("try to mining at height " + (best.height + 1))
            currentMiningHeight = best.height + 1
            val current = System.currentTimeMillis() / 1000
            if (current <= best.createdAt) return
        }
        val task = Runnable {
            try {
                poW.repo.reader.use { rd ->
                    val (block, infos) = createBlock(
                        rd,
                        rd.bestBlock
                    )
                    if (block != null) {
                        log.info("mining success block: {}", block.header)
                    }
                    eventBus.publish(NewBlockMined(block!!, infos))
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                task = null
            }
        }
        this.task = threadPool.submit(task)
    }

    companion object {
        private val log = LoggerFactory.getLogger("pow-miner")
    }

    init {
        eventBus.subscribe(NewBestBlock::class.java) { (block) ->
            if (block.height >= currentMiningHeight) {
                // cancel mining
                currentMiningHeight = 0
            }
        }
    }
}